File: java/src/main/java/com/ywh/problem/leetcode/medium/LeetCode36.java
package com.ywh.problem.leetcode.medium;

import java.util.HashSet;
import java.util.Set;

/**
 * 判断数独的有效性
 * [哈希表]
 *
 * @author ywh
 * @since 19/12/2019
 */
public class LeetCode36 {

    /**
     * Time: O(n^2), Space: O(1)
     *
     * @param board
     * @return
     */
    public boolean isValidSudokuBruteForce(char[][] board) {
        Set<Character> set = new HashSet<>();
        for (int i = 0; i < 9; i++) {

            // 判断纵列
            set.clear();
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                if (!set.add(board[i][j])) {
                    return false;
                }
            }

            // 判断横排
            set.clear();
            for (int j = 0; j < 9; j++) {
                if (board[j][i] == '.') {
                    continue;
                }
                if (!set.add(board[j][i])) {
                    return false;
                }
            }

            // 处理小正方形（i 表示第 i + 1 个正方形）
            set.clear();
            for (int j = 0; j < 9; j++) {
                int r = 3 * (i / 3) + j / 3;
                int c = 3 * (i % 3) + j % 3;
                if (board[r][c] == '.') {
                    continue;
                }
                if (!set.add(board[r][c])) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Time: O(n^2), Space: O(n^2)
     *
     * @param board
     * @return
     */
    public boolean isValidSudoku2DArray(char[][] board) {

        // i、j、k 表示小正方形的编号，x 表示 1 ~ 9 的数字
        boolean[][] rowSeen = new boolean[9][9];
        boolean[][] colSeen = new boolean[9][9];
        boolean[][] boxSeen = new boolean[9][9];
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int x = board[i][j] - '1';
                int k = (i / 3) * 3 + j / 3;
                if (rowSeen[i][x] || colSeen[j][x] || boxSeen[k][x]) {
                    return false;
                }
                // 表示在第 i 行中，x 已经出现过
                rowSeen[i][x] = true;
                colSeen[j][x] = true;
                boxSeen[k][x] = true;
            }
        }
        return true;
    }
}

File: python/src/problem/leetcode/medium/leetcode_36.py
# coding=utf-8
from typing import List


class Solution:
    """
    判断数独的有效性
    """

    def is_valid_sudoku(self, board: List[List[str]]) -> bool:
        """
        Time: O(n^2), Space: O(n^2)
        :param nums:
        :param target:
        :return:
        """

        for i in range(9):
            s = set()
            for j in range(9):
                if board[i][j] == ".":
                    continue
                if board[i][j] in s:
                    return False
                s.add(board[i][j])
            s = set()
            for j in range(9):
                if board[j][i] == ".":
                    continue
                if board[j][i] in s:
                    return False
                s.add(board[j][i])

            s = set()
            for j in range(9):
                r = 3 * (i // 3) + j // 3
                c = 3 * (i % 3) + j % 3
                if board[r][c] == ".":
                    continue
                if board[r][c] in s:
                    return False
                s.add(board[r][c])
        return True

    def is_valid_sudoku2(self, board: List[List[str]]) -> bool:
        """

        :param board:
        :return:
        """
        row_seen = [[False for _ in range(9)] for _ in range(9)]
        col_seen = [[False for _ in range(9)] for _ in range(9)]
        box_seen = [[False for _ in range(9)] for _ in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] == ".":
                    continue
                x = int(board[i][j]) - 1
                k = (i // 3) * 3 + j // 3
                if row_seen[i][x] or col_seen[j][x] or box_seen[k][x]:
                    return False
                row_seen[i][x] = col_seen[j][x] = box_seen[k][x] = True
        return True

